sklearn/metrics/_scorer.py

```diff
@@ -130,7 +129,21 @@ def __call__(self, estimator, *args, **kwargs):
                 **{name: Bunch(score=kwargs) for name in self._scorers}
             )
 
-        for name, scorer in self._scorers.items():
+        # to have the highest cache hit rate, we need to check if we have any
+        # _ThresholdScorer and choose a single response method based on the current
+        # estimator
+        scorers = copy.deepcopy(self._scorers)
```

Comment:
Why do we need to deepcopy the `_scorers` dict? My two initial reactions when seeing `deepcopy` are: (1) surely this isn't needed because you need the `copy` module very rarely and (2) you better explain why you need to do something weird like using `deepcopy`

Reply:
We never mutate arguments and since `check_scoring` is not making a copy, we would mutate the argument passed by the user here. This is the reason for making a deep copy.

Reply:
I feel like there is more copying than there needs to be. I think we can do:

```python
def process_scorer(scorer):
    if not isinstance(scorer, _ThresholdScorer):
        return scorer
    if scorer._response_method is None:
        response_method = ("decision_function", "predict_proba")
    else:
        response_method = scorer._response_method
    new_response_method = _check_response_method(
        estimator, response_method).__name__

    # no need for deep copy
    new_scorer = scorer.copy()
    new_scorer._response_method = new_response_method
    return new_scorer

scorers = {name: process_scorer(scorer) name, scorer in self._scorers.items():
```